tsl/src/compression/compression.c

```diff
@@ -2196,47 +2199,51 @@ compressed_insert_key_columns(Relation relation)
 	if (indexoidlist == NIL)
 		return NULL;
 
-	/*
-	 * For each index, add referenced attributes to indexattrs.
-	 *
-	 * Note: we consider all indexes returned by RelationGetIndexList, even if
-	 * they are not indisready or indisvalid.  This is important because an
-	 * index for which CREATE INDEX CONCURRENTLY has just started must be
-	 * included in HOT-safety decisions (see README.HOT).  If a DROP INDEX
-	 * CONCURRENTLY is far enough along that we should ignore the index, it
-	 * won't be returned at all by RelationGetIndexList.
-	 */
 	foreach (l, indexoidlist)
 	{
 		Oid indexOid = lfirst_oid(l);
 		Relation indexDesc = index_open(indexOid, AccessShareLock);
 
-		if (!indexDesc->rd_index->indisunique)
+		/*
+		 * We are only interested in unique indexes. PRIMARY KEY indexes also have
+		 * indisunique set to true so we do not need to check for them separately.
+		 */
+		if (!indexDesc->rd_index->indislive || !indexDesc->rd_index->indisvalid ||
+			!indexDesc->rd_index->indisunique)
 		{
 			index_close(indexDesc, AccessShareLock);
 			continue;
 		}
 
-		/* Collect simple attribute references.
-		 * For covering indexes we only need to collect the key attributes.
-		 * Unlike RelationGetIndexAttrBitmap we allow expression indexes
-		 * but we do not extract attributes from the expressions as that
-		 * would not be a safe filter as the expression can alter attributes
-		 * which would not make them sufficient for batch filtering.
+		Bitmapset *idx_attrs = NULL;
+		/*
+		 * Collect attributes of current index.
+		 * For covering indexes we need to ignore the included columns.
 		 */
 		for (int i = 0; i < indexDesc->rd_index->indnkeyatts; i++)
 		{
 			int attrnum = indexDesc->rd_index->indkey.values[i];
-			if (attrnum != 0)
-			{
-				indexattrs =
-					bms_add_member(indexattrs, attrnum - FirstLowInvalidHeapAttributeNumber);
-			}
+			/* We are not interested in expression columns which will have attrnum = 0 */
+			if (!attrnum)
+				continue;
+
+			idx_attrs = bms_add_member(idx_attrs, attrnum - FirstLowInvalidHeapAttributeNumber);
 		}
 		index_close(indexDesc, AccessShareLock);
+
+		if (first_iteration)
```

Comment:
```suggestion
		if (!shared_attrs)
```

Reply:
that wouldnt handle the case when we have non-intersecting indexes resetting shared_atts